File: src/utils/excelAnalyzer.ts
/**
 * Excel File Analyzer Utility
 * Analyzes Excel files to extract column structure and sample data
 * This helps map real Excel columns to TypeScript interfaces
 */

import * as XLSX from 'xlsx';

export interface ColumnInfo {
  name: string;
  index: number;
  type: 'string' | 'number' | 'date' | 'boolean' | 'unknown';
  sampleValues: unknown[];
  uniqueCount: number;
  nullCount: number;
}

export interface ExcelAnalysisResult {
  fileName: string;
  sheetName: string;
  totalRows: number;
  totalColumns: number;
  columns: ColumnInfo[];
  sampleRows: Record<string, unknown>[];
  recommendations: {
    artikelnummer: string[];
    projektnummer: string[];
    dates: string[];
    quantities: string[];
    status: string[];
    other: string[];
  };
  businessRules?: {
    filterColumn?: string;
    filterCondition?: string;
    estimatedRowsToSkip?: number;
  };
}

/**
 * Keywords for identifying specific column types
 */
const COLUMN_KEYWORDS = {
  artikelnummer: ['art', 'item', 'artikel', 'material', 'nummer', 'itemnr', 'itemno', 'artnr'],
  projektnummer: ['proj', 'project', 'auftrag', 'order', 'projektnr', 'projnr'],
  date: ['date', 'datum', 'termin', 'deadline', 'start', 'end', 'ende'],
  quantity: ['qty', 'quantity', 'menge', 'anzahl', 'rem', 'remaining'],
  status: ['status', 'state', 'zustand'],
  customer: ['customer', 'kunde', 'client'],
  description: ['description', 'beschreibung', 'name', 'bezeichnung'],
  price: ['price', 'preis', 'cost', 'kosten'],
};

/**
 * Analyze an Excel file from File object
 */
export async function analyzeExcelFile(
  file: File,
  options: {
    maxSampleRows?: number;
    maxSampleValues?: number;
  } = {}
): Promise<ExcelAnalysisResult> {
  const { maxSampleRows = 5, maxSampleValues = 3 } = options;

  // Read file as ArrayBuffer
  const arrayBuffer = await file.arrayBuffer();

  // Read workbook
  const workbook = XLSX.read(arrayBuffer, {
    type: 'array',
    cellDates: true,
  });

  // Get first sheet
  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];

  // Convert to JSON
  const jsonData = XLSX.utils.sheet_to_json<Record<string, unknown>>(sheet, {
    raw: false,
    defval: null,
  });

  // Remove completely empty rows
  const rows = jsonData.filter((row) =>
    Object.values(row).some((val) => val !== null && val !== '')
  );

  const totalRows = rows.length;
  const columnNames = rows.length > 0 ? Object.keys(rows[0]) : [];
  const totalColumns = columnNames.length;

  // Analyze each column
  const columns: ColumnInfo[] = columnNames.map((columnName, index) => {
    const columnValues = rows.map((row) => row[columnName]);

    // Get sample values (non-null)
    const nonNullValues = columnValues.filter(
      (v) => v !== null && v !== undefined && v !== ''
    );
    const sampleValues = nonNullValues.slice(0, maxSampleValues);

    // Count unique and null values
    const uniqueCount = new Set(nonNullValues).size;
    const nullCount = columnValues.length - nonNullValues.length;

    return {
      name: columnName,
      index,
      type: detectColumnType(columnValues),
      sampleValues,
      uniqueCount,
      nullCount,
    };
  });

  // Get sample rows
  const sampleRows = rows.slice(0, maxSampleRows);

  // Generate recommendations
  const recommendations = generateRecommendations(columns);

  // Check for business rules (e.g., QuantityRem1 filtering)
  const businessRules = detectBusinessRules(columns, rows);

  return {
    fileName: file.name,
    sheetName,
    totalRows,
    totalColumns,
    columns,
    sampleRows,
    recommendations,
    businessRules,
  };
}

/**
 * Analyze Excel file from path (Node.js environment)
 */
export function analyzeExcelFileFromPath(
  filePath: string,
  options: {
    maxSampleRows?: number;
    maxSampleValues?: number;
  } = {}
): ExcelAnalysisResult {
  const { maxSampleRows = 5, maxSampleValues = 3 } = options;

  // Read workbook
  const workbook = XLSX.readFile(filePath, {
    cellDates: true,
  });

  // Get first sheet
  const sheetName = workbook.SheetNames[0];
  const sheet = workbook.Sheets[sheetName];

  // Convert to JSON
  const jsonData = XLSX.utils.sheet_to_json<Record<string, unknown>>(sheet, {
    raw: false,
    defval: null,
  });

  // Remove completely empty rows
  const rows = jsonData.filter((row) =>
    Object.values(row).some((val) => val !== null && val !== '')
  );

  const totalRows = rows.length;
  const columnNames = rows.length > 0 ? Object.keys(rows[0]) : [];
  const totalColumns = columnNames.length;

  // Analyze each column
  const columns: ColumnInfo[] = columnNames.map((columnName, index) => {
    const columnValues = rows.map((row) => row[columnName]);

    // Get sample values (non-null)
    const nonNullValues = columnValues.filter(
      (v) => v !== null && v !== undefined && v !== ''
    );
    const sampleValues = nonNullValues.slice(0, maxSampleValues);

    // Count unique and null values
    const uniqueCount = new Set(nonNullValues).size;
    const nullCount = columnValues.length - nonNullValues.length;

    return {
      name: columnName,
      index,
      type: detectColumnType(columnValues),
      sampleValues,
      uniqueCount,
      nullCount,
    };
  });

  // Get sample rows
  const sampleRows = rows.slice(0, maxSampleRows);

  // Generate recommendations
  const recommendations = generateRecommendations(columns);

  // Check for business rules
  const businessRules = detectBusinessRules(columns, rows);

  const fileName = filePath.split('/').pop() || filePath;

  return {
    fileName,
    sheetName,
    totalRows,
    totalColumns,
    columns,
    sampleRows,
    recommendations,
    businessRules,
  };
}

/**
 * Detect column type based on values
 */
function detectColumnType(
  values: unknown[]
): 'string' | 'number' | 'date' | 'boolean' | 'unknown' {
  const sampleSize = Math.min(100, values.length);
  const samples = values.slice(0, sampleSize);

  let numberCount = 0;
  let stringCount = 0;
  let dateCount = 0;
  let booleanCount = 0;

  for (const value of samples) {
    if (value === null || value === undefined || value === '') {
      continue;
    }

    if (typeof value === 'boolean') {
      booleanCount++;
    } else if (typeof value === 'number') {
      numberCount++;
    } else if (value instanceof Date) {
      dateCount++;
    } else if (typeof value === 'string') {
      if (isDateString(value)) {
        dateCount++;
      } else {
        stringCount++;
      }
    }
  }

  const total = numberCount + stringCount + dateCount + booleanCount;
  if (total === 0) return 'unknown';

  // Determine dominant type
  if (dateCount / total > 0.7) return 'date';
  if (numberCount / total > 0.7) return 'number';
  if (booleanCount / total > 0.7) return 'boolean';
  if (stringCount / total > 0.5) return 'string';

  return 'unknown';
}

/**
 * Check if string looks like a date
 */
function isDateString(str: string): boolean {
  const datePatterns = [
    /^\d{4}-\d{2}-\d{2}/, // YYYY-MM-DD
    /^\d{2}\.\d{2}\.\d{4}/, // DD.MM.YYYY
    /^\d{2}\/\d{2}\/\d{4}/, // DD/MM/YYYY or MM/DD/YYYY
  ];

  return datePatterns.some((pattern) => pattern.test(str));
}

/**
 * Generate recommendations for column mappings
 */
function generateRecommendations(columns: ColumnInfo[]) {
  const recommendations = {
    artikelnummer: [] as string[],
    projektnummer: [] as string[],
    dates: [] as string[],
    quantities: [] as string[],
    status: [] as string[],
    other: [] as string[],
  };

  for (const column of columns) {
    const lowerName = column.name.toLowerCase();

    // Check for Artikelnummer
    if (COLUMN_KEYWORDS.artikelnummer.some((kw) => lowerName.includes(kw))) {
      recommendations.artikelnummer.push(column.name);
      continue;
    }

    // Check for Projektnummer
    if (COLUMN_KEYWORDS.projektnummer.some((kw) => lowerName.includes(kw))) {
      recommendations.projektnummer.push(column.name);
      continue;
    }

    // Check for dates
    if (COLUMN_KEYWORDS.date.some((kw) => lowerName.includes(kw)) || column.type === 'date') {
      recommendations.dates.push(column.name);
      continue;
    }

    // Check for quantities
    if (COLUMN_KEYWORDS.quantity.some((kw) => lowerName.includes(kw))) {
      recommendations.quantities.push(column.name);
      continue;
    }

    // Check for status
    if (COLUMN_KEYWORDS.status.some((kw) => lowerName.includes(kw))) {
      recommendations.status.push(column.name);
      continue;
    }

    // Other important columns
    if (
      COLUMN_KEYWORDS.customer.some((kw) => lowerName.includes(kw)) ||
      COLUMN_KEYWORDS.description.some((kw) => lowerName.includes(kw)) ||
      COLUMN_KEYWORDS.price.some((kw) => lowerName.includes(kw))
    ) {
      recommendations.other.push(column.name);
    }
  }

  return recommendations;
}

/**
 * Detect business rules based on column analysis
 */
function detectBusinessRules(
  columns: ColumnInfo[],
  rows: Record<string, unknown>[]
): {
  filterColumn?: string;
  filterCondition?: string;
  estimatedRowsToSkip?: number;
} | undefined {
  // Check for QuantityRem1 (Sales file business rule)
  const qtyRemCol = columns.find((col) => col.name === 'QuantityRem1');

  if (qtyRemCol) {
    const zeroCount = rows.filter((row) => row['QuantityRem1'] === 0).length;

    return {
      filterColumn: 'QuantityRem1',
      filterCondition: 'Skip rows where QuantityRem1 == 0 (already delivered)',
      estimatedRowsToSkip: zeroCount,
    };
  }

  return undefined;
}

/**
 * Format analysis result as readable text
 */
export function formatAnalysisResult(result: ExcelAnalysisResult): string {
  const lines: string[] = [];

  lines.push('='.repeat(100));
  lines.push(`FILE: ${result.fileName}`);
  lines.push(`Sheet: ${result.sheetName}`);
  lines.push('='.repeat(100));
  lines.push('');

  lines.push('BASIC INFORMATION:');
  lines.push(`  Total rows: ${result.totalRows}`);
  lines.push(`  Total columns: ${result.totalColumns}`);
  lines.push('');

  lines.push(`ALL COLUMN NAMES (${result.columns.length} columns):`);
  result.columns.forEach((col, idx) => {
    lines.push(`  ${String(idx + 1).padStart(2, ' ')}. ${col.name} (${col.type})`);
    if (col.sampleValues.length > 0) {
      const samples = col.sampleValues.map((v) => String(v).substring(0, 30)).join(', ');
      lines.push(`      Samples: ${samples}`);
    }
  });
  lines.push('');

  lines.push('COLUMN RECOMMENDATIONS:');
  lines.push('-'.repeat(100));

  if (result.recommendations.artikelnummer.length > 0) {
    lines.push(`  ARTIKELNUMMER (Article Number): ${result.recommendations.artikelnummer.join(', ')}`);
  } else {
    lines.push('  ARTIKELNUMMER (Article Number): NOT FOUND - manual mapping needed');
  }

  if (result.recommendations.projektnummer.length > 0) {
    lines.push(`  PROJEKTNUMMER (Project Number): ${result.recommendations.projektnummer.join(', ')}`);
  } else {
    lines.push('  PROJEKTNUMMER (Project Number): NOT FOUND - manual mapping needed');
  }

  if (result.recommendations.dates.length > 0) {
    lines.push(`  DATE columns: ${result.recommendations.dates.join(', ')}`);
  }

  if (result.recommendations.quantities.length > 0) {
    lines.push(`  QUANTITY columns: ${result.recommendations.quantities.join(', ')}`);
  }

  if (result.recommendations.status.length > 0) {
    lines.push(`  STATUS columns: ${result.recommendations.status.join(', ')}`);
  }

  if (result.recommendations.other.length > 0) {
    lines.push(`  OTHER important columns: ${result.recommendations.other.join(', ')}`);
  }

  lines.push('');

  if (result.businessRules) {
    lines.push('BUSINESS RULES:');
    lines.push('*'.repeat(100));
    lines.push(`  Filter Column: ${result.businessRules.filterColumn}`);
    lines.push(`  Condition: ${result.businessRules.filterCondition}`);
    lines.push(`  Estimated rows to skip: ${result.businessRules.estimatedRowsToSkip} of ${result.totalRows} (${((result.businessRules.estimatedRowsToSkip || 0) / result.totalRows * 100).toFixed(1)}%)`);
    lines.push('*'.repeat(100));
    lines.push('');
  }

  lines.push('SAMPLE DATA (first few rows):');
  lines.push('-'.repeat(100));
  result.sampleRows.forEach((row, idx) => {
    lines.push(`\nRow ${idx + 1}:`);
    Object.entries(row).forEach(([key, value]) => {
      if (value !== null && value !== undefined && value !== '') {
        const valueStr = String(value).substring(0, 60);
        lines.push(`  ${key}: ${valueStr}`);
      }
    });
  });

  lines.push('');
  lines.push('='.repeat(100));

  return lines.join('\n');
}

File: excel_analyzer.py
#!/usr/bin/env python3
"""
Excel File Analyzer for ERP Data
Analyzes three Excel files and extracts column structure and sample data
"""

import sys
import os

# Try to import required libraries
try:
    import pandas as pd
    use_pandas = True
    print("Using pandas for analysis")
except ImportError:
    use_pandas = False
    print("Pandas not available, trying openpyxl...")

if not use_pandas:
    try:
        from openpyxl import load_workbook
        use_openpyxl = True
        print("Using openpyxl for analysis")
    except ImportError:
        print("ERROR: Neither pandas nor openpyxl is available!")
        print("Please install one of them:")
        print("  pip install pandas openpyxl")
        print("  or")
        print("  pip install openpyxl")
        sys.exit(1)

# File paths
FILES = {
    'Sales (Offene Lieferungen)': '/Users/phillipplomer/Desktop/Programme/Neubau Claude/2025-09-30_Offene_Lieferungen_Stand.xlsx',
    'Production Planning': '/Users/phillipplomer/Desktop/Programme/Neubau Claude/2025-12-10_PP_SollIstVergleich.xlsm',
    'Project Management (Controlling)': '/Users/phillipplomer/Desktop/Programme/Neubau Claude/Controlling.xlsx'
}


def analyze_with_pandas(filepath, name):
    """Analyze Excel file using pandas"""
    print(f"\n{'='*100}")
    print(f"FILE: {name}")
    print(f"Path: {filepath}")
    print('='*100)

    try:
        # Read Excel file
        df = pd.read_excel(filepath, engine='openpyxl')

        # Basic information
        print(f"\nBASIC INFORMATION:")
        print(f"  Total rows (including header): {len(df) + 1}")
        print(f"  Data rows: {len(df)}")
        print(f"  Total columns: {len(df.columns)}")

        # All column names
        print(f"\nALL COLUMN NAMES ({len(df.columns)} columns):")
        for i, col in enumerate(df.columns, 1):
            print(f"  {i:2d}. {col}")

        # Sample data - first 5 rows
        print(f"\nSAMPLE DATA (First 5 rows):")
        print("-" * 100)

        # Show data in a readable format
        for idx in range(min(5, len(df))):
            print(f"\nRow {idx + 1}:")
            for col in df.columns:
                value = df.iloc[idx][col]
                if pd.notna(value):  # Only show non-null values
                    # Format value
                    if isinstance(value, float):
                        value_str = f"{value:.2f}" if value != int(value) else f"{int(value)}"
                    else:
                        value_str = str(value)

                    # Truncate long values
                    if len(value_str) > 60:
                        value_str = value_str[:60] + "..."

                    print(f"  {col}: {value_str}")

        # Column Analysis
        print(f"\n\nCOLUMN ANALYSIS:")
        print("-" * 100)

        # Article Number columns
        article_candidates = [col for col in df.columns
                            if any(x in str(col).lower() for x in ['art', 'item', 'artikel', 'material', 'nummer'])]
        if article_candidates:
            print(f"\nPotential ARTICLE NUMBER columns:")
            for col in article_candidates:
                sample_val = df[col].dropna().iloc[0] if len(df[col].dropna()) > 0 else "N/A"
                print(f"  - {col}")
                print(f"    Sample value: {sample_val}")
                print(f"    Unique values: {df[col].nunique()}")

        # Project Number columns
        project_candidates = [col for col in df.columns
                            if any(x in str(col).lower() for x in ['proj', 'project', 'auftrag', 'order'])]
        if project_candidates:
            print(f"\nPotential PROJECT NUMBER columns:")
            for col in project_candidates:
                sample_val = df[col].dropna().iloc[0] if len(df[col].dropna()) > 0 else "N/A"
                print(f"  - {col}")
                print(f"    Sample value: {sample_val}")
                print(f"    Unique values: {df[col].nunique()}")

        # Quantity columns
        qty_candidates = [col for col in df.columns
                        if any(x in str(col).lower() for x in ['qty', 'quantity', 'menge', 'anzahl', 'rem'])]
        if qty_candidates:
            print(f"\nPotential QUANTITY columns:")
            for col in qty_candidates:
                if pd.api.types.is_numeric_dtype(df[col]):
                    print(f"  - {col}")
                    print(f"    Min: {df[col].min()}, Max: {df[col].max()}, Mean: {df[col].mean():.2f}")

        # Date columns
        date_candidates = [col for col in df.columns
                         if any(x in str(col).lower() for x in ['date', 'datum', 'termin', 'deadline'])]
        if date_candidates:
            print(f"\nPotential DATE columns:")
            for col in date_candidates:
                sample_val = df[col].dropna().iloc[0] if len(df[col].dropna()) > 0 else "N/A"
                print(f"  - {col}")
                print(f"    Sample value: {sample_val}")

        # Status columns
        status_candidates = [col for col in df.columns
                           if any(x in str(col).lower() for x in ['status', 'state', 'zustand'])]
        if status_candidates:
            print(f"\nPotential STATUS columns:")
            for col in status_candidates:
                unique_vals = df[col].unique()[:5]  # First 5 unique values
                print(f"  - {col}")
                print(f"    Unique values (first 5): {unique_vals}")

        # Special handling for Sales file
        if 'QuantityRem1' in df.columns:
            print(f"\n{'*'*100}")
            print(f"IMPORTANT: QuantityRem1 column found (BUSINESS RULE)")
            print(f"{'*'*100}")
            zero_count = len(df[df['QuantityRem1'] == 0])
            non_zero_count = len(df[df['QuantityRem1'] > 0])
            total = len(df)
            print(f"  Total rows: {total}")
            print(f"  Rows where QuantityRem1 = 0 (already delivered, SKIP): {zero_count} ({zero_count/total*100:.1f}%)")
            print(f"  Rows where QuantityRem1 > 0 (pending delivery, INCLUDE): {non_zero_count} ({non_zero_count/total*100:.1f}%)")
            print(f"\n  ACTION: Filter out rows where QuantityRem1 == 0 before processing")

        # Recommendations
        print(f"\n\nRECOMMENDATIONS:")
        print("-" * 100)

        if article_candidates:
            print(f"  ARTIKELNUMMER (Article Number): Use column '{article_candidates[0]}'")
        else:
            print(f"  ARTIKELNUMMER (Article Number): Not clearly identified - manual review needed")

        if project_candidates:
            print(f"  PROJEKTNUMMER (Project Number): Use column '{project_candidates[0]}'")
        else:
            print(f"  PROJEKTNUMMER (Project Number): Not clearly identified - manual review needed")

    except Exception as e:
        print(f"\nERROR analyzing file: {str(e)}")
        import traceback
        traceback.print_exc()


def analyze_with_openpyxl(filepath, name):
    """Analyze Excel file using openpyxl (fallback method)"""
    print(f"\n{'='*100}")
    print(f"FILE: {name}")
    print(f"Path: {filepath}")
    print('='*100)

    try:
        # Load workbook
        wb = load_workbook(filepath, read_only=True, data_only=True)
        sheet = wb.active

        print(f"\nBASIC INFORMATION:")
        print(f"  Sheet name: {sheet.title}")
        print(f"  Total rows: {sheet.max_row}")
        print(f"  Total columns: {sheet.max_column}")

        # Get column names
        column_names = []
        for col in range(1, sheet.max_column + 1):
            cell_value = sheet.cell(row=1, column=col).value
            column_names.append(cell_value if cell_value is not None else f"Column_{col}")

        # Display column names
        print(f"\nALL COLUMN NAMES ({len(column_names)} columns):")
        for i, col in enumerate(column_names, 1):
            print(f"  {i:2d}. {col}")

        # Display sample data
        print(f"\nSAMPLE DATA (First 5 rows):")
        print("-" * 100)

        max_rows = min(6, sheet.max_row)
        for row in range(2, max_rows + 1):
            print(f"\nRow {row - 1}:")
            for col_idx, col_name in enumerate(column_names, 1):
                cell_value = sheet.cell(row=row, column=col_idx).value
                if cell_value is not None:
                    value_str = str(cell_value)
                    if len(value_str) > 60:
                        value_str = value_str[:60] + "..."
                    print(f"  {col_name}: {value_str}")

        # Column analysis
        print(f"\n\nCOLUMN ANALYSIS:")
        print("-" * 100)

        article_candidates = [col for col in column_names
                            if col and any(x in str(col).lower() for x in ['art', 'item', 'artikel', 'material'])]
        project_candidates = [col for col in column_names
                            if col and any(x in str(col).lower() for x in ['proj', 'project', 'auftrag'])]
        qty_candidates = [col for col in column_names
                        if col and any(x in str(col).lower() for x in ['qty', 'quantity', 'menge', 'anzahl', 'rem'])]
        date_candidates = [col for col in column_names
                         if col and any(x in str(col).lower() for x in ['date', 'datum', 'termin'])]

        if article_candidates:
            print(f"\nPotential ARTICLE NUMBER columns: {article_candidates}")
        if project_candidates:
            print(f"\nPotential PROJECT NUMBER columns: {project_candidates}")
        if qty_candidates:
            print(f"\nPotential QUANTITY columns: {qty_candidates}")
        if date_candidates:
            print(f"\nPotential DATE columns: {date_candidates}")

        # Check for QuantityRem1
        if 'QuantityRem1' in column_names:
            print(f"\n{'*'*100}")
            print(f"IMPORTANT: QuantityRem1 column found")
            print(f"  BUSINESS RULE: Rows with QuantityRem1 = 0 should be skipped (already delivered)")
            print(f"{'*'*100}")

        wb.close()

    except Exception as e:
        print(f"\nERROR analyzing file: {str(e)}")
        import traceback
        traceback.print_exc()


def main():
    """Main analysis function"""
    print("="*100)
    print("EXCEL FILE ANALYZER - ERP Data Analysis")
    print("="*100)

    # Check files exist
    print("\nChecking files...")
    for name, filepath in FILES.items():
        exists = "EXISTS" if os.path.exists(filepath) else "NOT FOUND"
        print(f"  {name}: {exists}")

    print("\n")

    # Analyze each file
    for name, filepath in FILES.items():
        if not os.path.exists(filepath):
            print(f"\nSKIPPING {name}: File not found")
            continue

        if use_pandas:
            analyze_with_pandas(filepath, name)
        else:
            analyze_with_openpyxl(filepath, name)

    print("\n" + "="*100)
    print("ANALYSIS COMPLETE")
    print("="*100)


if __name__ == "__main__":
    main()
